### ISSUE TITLE
Regent Stencil results on Perlmutter

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1318]
This is the sister issue to #1270 for Perlmutter. I am running two configurations, MPI (`REALM_NETWORKS=mpi`) and GASNet (`REALM_NETWORKS=gasnet`). All runs use the following variable (and no other relevant variables):

```
export FI_CXI_DEFAULT_CQ_SIZE=1310720
```

Note that Perlmutter is now fully upgraded to Slingshot 11, the same network as Crusher (and Frontier).

This uses the latest fixes from #1316, so startup time should be dramatically reduced.

## MPI Results (Weak Scaling)

MPI results go up to 512 nodes. Time to first iteration is mostly linear until about 256 nodes where it's clear we're starting to see a superlinear term (not quite quadratic) emerge. We're also seeing a drop-off in performance starting at 256 nodes.

nodes | time to first iteration (seconds) | remaining time (seconds)
-- | -- | --
1 | 0.09275 | 1.684
2 | 0.1115 | 1.7045
4 | 0.14025 | 1.711
8 | 0.2005 | 1.7255
16 | 0.3165 | 1.757
32 | 0.559 | 1.8685
64 | 1.004 | 1.915
128 | 2.115666667 | 2.234333333
256 | 4.9245 | 3.4825
512 | 13.4085 | 6.5635

Note that "time to first iteration" begins at the start of the top-level task. At higher node counts, there is a fair bit of startup time that goes on before this, in SLURM, Legion and/or Realm initialization, which is noticeable in the difference between these measured times and the total execution time for the batch script.

## GASNet Results (Weak Scaling)

GASNet results seem to be less well-behaved for some reason. We're getting OOM-killed at 512 nodes and the 64-node run appeared to freeze (a shutdown hang, meaning that we got through one run but timed out the job).

nodes | time to first iteration (seconds) | remaining time (seconds)
-- | -- | --
1 | 0.07875 | 1.68575
2 | 0.084 | 1.695
4 | 0.0925 | 1.7035
8 | 0.1115 | 1.71325
16 | 0.14675 | 1.72775
32 | 0.23 | 1.7685
64 | Freeze? | Freeze?
128 | 0.86525 | 1.80175
256 | 2.226 | 1.926666667
512 | OOM | OOM

The OOM error looks like this:

```
slurmstepd: error: Detected 1 oom-kill event(s) in StepId=3097023.0. Some of your processes may have been killed by the cgroup out-of-memory
 handler.
srun: error: nid001049: tasks 1-7: Out Of Memory
```

There are backtraces beyond this, but I don't think they're useful, since they refer to after the point where GASNet is already being torn down.

## Analysis

Compared to the most recent status of #1270, we were able to get to 512 nodes (4096 ranks) for Perlmutter, a size we're unable to test on Crusher do to the limited node count on that machine. However it seems that we're hitting superlinear startup time (again) and GASNet is not entirely behaving itself.

**Comment by lightsighter** [comment]
Do you have profiles for any of these?

**Comment by elliottslaughter** [comment]
Here are some profiles.

MPI:

  * 64 nodes: http://sapling.stanford.edu/~eslaught/perlmutter_startup_scaling_2022-09-06/mpi/legion_prof_64x1_r0
  * 128 nodes: http://sapling.stanford.edu/~eslaught/perlmutter_startup_scaling_2022-09-06/mpi/legion_prof_128x1_r0
  * 256 nodes: http://sapling.stanford.edu/~eslaught/perlmutter_startup_scaling_2022-09-06/mpi/legion_prof_256x1_r0

GASNet:

  * 64 nodes: http://sapling.stanford.edu/~eslaught/perlmutter_startup_scaling_2022-09-06/gasnet/legion_prof_64x1_r0
  * 128 nodes: timed out
  * 256 nodes: http://sapling.stanford.edu/~eslaught/perlmutter_startup_scaling_2022-09-06/gasnet/legion_prof_256x1_r0

**Comment by elliottslaughter** [comment]
I can't get the GASNet 256 node profile to open in either Firefox or Chrome. Here's a version with node 0 excluded so at least we can see something: http://sapling.stanford.edu/~eslaught/perlmutter_startup_scaling_2022-09-06/gasnet/legion_prof_256x1_r0_node1plus

Here's what I notice in the profiles:

  * In MPI from 64 to 128 nodes there is a dramatic increase in the amount of time it takes to get to the start of the top-level task: about 12 seconds to 1,500 seconds. This is a factor of 150&times for only 2&times; more nodes!
  * The difference is less dramatic in GASNet: from 10 seconds (64 nodes) to 240 seconds (256 nodes), a factor of 24&times; for 4&times; more nodes (still worse than quadratic).
  * In both GASNet and MPI, *after the application completes*, I see a huge amount of activity on node 0's utility processor: `Send View Find Copy Preconditions Request`, initiated by `get_elapsed`. I think that's this task: https://gitlab.com/StanfordLegion/legion/-/blob/crusher-freeze-repro-perlmutter-2-gasnet/language/examples/stencil_fast.rg#L551 . This is not too surprising because it's sucking a (small) region onto one node. I'll try to find a way to do this with futures instead.
  * However, I do not see an obvious root cause for why the "time to first iteration" (i.e., from the start of the top-level task to the first completed operation) would be growing quadratically. Maybe it's just hard to see relative to the other factors right now.

So in summary it looks like one of the two issues may be the application's fault, but the other seems likely to be runtime startup, and I don't have an answer to quadratic time to first iteration yet.

**Comment by elliottslaughter** [comment]
Please disregard these profiles. I remembered that the first run in a job usually has an abnormally high startup time---probably the result of the filesystem.

I'll fix this (and the application issue) and collect a new set of logs.
